requirements/progress.py

```diff
@@ -0,0 +1,318 @@
+            manual_progress = progress_overrides[self.list_path]
+        else:
+            manual_progress = 0
+        if self.statement.requirement is not None:
+            #it is a basic requirement
+            if self.statement.is_plain_string and not manual_progress == 0 and self.statement.threshold_type is not None:
+                #use manual progress
+                is_fulfilled = manual_progress >= current_threshold.get_actual_cutoff()
+                subjects = 0
+                units = 0
+                if current_threshold.criterion == CRITERION_UNITS:
+                    units = manual_progress
+                    subjects = manual_progress / DEFAULT_UNIT_COUNT
+                else:
+                    units = manual_progress * DEFAULT_UNIT_COUNT
+                    subjects = manual_progress
+                subject_progress = ceiling_thresh(subjects, current_threshold.cutoff_for_criterion(CRITERION_SUBJECTS))
+                unit_progress = ceiling_thresh(units, current_threshold.cutoff_for_criterion(CRITERION_UNITS))
+                #fill with dummy courses
+                random_ids = random.sample(range(1000, max(10000, subject_progress.progress+1000)), subject_progress.progress)
+                for rand_id in random_ids:
+                    dummy_course = Course(id = self.list_path + "_"+str(rand_id), subject_id="gen_course_"+self.list_path+"_"+str(rand_id),title="Generated Course "+self.list_path + " " + str(rand_id))
+                    satisfied_courses.add(dummy_course)
+            else:
+                #Example: requirement CI-H, we want to show how many have been fulfilled
+                satisfied_courses = self.courses_satisfying_req(courses)
+                if not self.statement.threshold_type is None:
+                    #A specific number of courses is required
+                    subject_progress = ceiling_thresh(len(satisfied_courses), current_threshold.cutoff_for_criterion(CRITERION_SUBJECTS))
+                    unit_progress = ceiling_thresh(total_units(satisfied_courses), current_threshold.cutoff_for_criterion(CRITERION_UNITS))
+                    is_fulfilled = current_threshold.is_satisfied_by(subject_progress.progress, unit_progress.progress)
+                else:
+                    #Only one is needed
+                    progress_subjects = min(len(satisfied_courses), 1)
+                    is_fulfilled = len(satisfied_courses) > 0
+                    subject_progress = ceiling_thresh(progress_subjects, 1)
+                    if len(satisfied_courses) > 0:
+                        unit_progress = ceiling_thresh(list(satisfied_courses)[0].total_units, DEFAULT_UNIT_COUNT)
+                    else:
+                        unit_progress = ceiling_thresh(0, DEFAULT_UNIT_COUNT)
+            progress = (subject_progress, unit_progress)[self.statement.threshold_type is not None and self.statement.threshold_criterion == CRITERION_UNITS]
```

Comment:
Ack, this is some weird syntax.
```suggestion
            progress = unit_progress if (current_threshold.type is not None and current_threshold.criterion == CRITERION_UNITS) else subject_progress
```